track_analyzing/serialization.hpp

```diff
@@ -0,0 +1,175 @@
+#pragma once
+
+#include "track_analyzing/track.hpp"
+
+#include "routing/num_mwm_id.hpp"
+#include "routing/segment.hpp"
+
+#include "platform/country_file.hpp"
+
+#include "coding/read_write_utils.hpp"
+#include "coding/reader.hpp"
+#include "coding/traffic.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+#include "coding/writer.hpp"
+
+#include "base/assert.hpp"
+#include "base/checked_cast.hpp"
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+namespace tracking
+{
+class MwmToMatchedTracksSerializer final
+{
+public:
+  MwmToMatchedTracksSerializer(std::shared_ptr<routing::NumMwmIds> numMwmIds)
+    : m_numMwmIds(numMwmIds)
+  {
+  }
+
+  template <class Sink>
+  void Serialize(MwmToMatchedTracks const & mwmToMatchedTracks, Sink & sink)
+  {
+    WriteSize(sink, mwmToMatchedTracks.size());
+
+    for (auto const & mwmIt : mwmToMatchedTracks)
+    {
+      rw::Write(sink, m_numMwmIds->GetFile(mwmIt.first).GetName());
+
+      UserToMatchedTracks const & userToMatchedTracks = mwmIt.second;
+      CHECK(!userToMatchedTracks.empty(), ());
+      WriteSize(sink, userToMatchedTracks.size());
+
+      for (auto const & userIt : userToMatchedTracks)
+      {
+        rw::Write(sink, userIt.first);
+
+        std::vector<MatchedTrack> const & tracks = userIt.second;
+        CHECK(!tracks.empty(), ());
+        WriteSize(sink, tracks.size());
+
+        for (MatchedTrack const & track : tracks)
+        {
+          CHECK(!track.empty(), ());
+          WriteSize(sink, track.size());
+
+          for (MatchedTrackPoint const & point : track)
+            Serialize(point.GetSegment(), sink);
+
+          std::vector<DataPoint> dataPoints;
+          dataPoints.reserve(track.size());
+          for (MatchedTrackPoint const & point : track)
+            dataPoints.push_back(point.GetDataPoint());
+
+          std::vector<uint8_t> buffer;
+          MemWriter<decltype(buffer)> memWriter(buffer);
+          coding::TrafficGPSEncoder::SerializeDataPoints(coding::TrafficGPSEncoder::kLatestVersion,
```

Comment:
Up to you, but generally it's a good idea for encoders/decoders to support iterators or iterator ranges.